Image: Add lightbox using directives.

artemiomorales opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

*Updated testing instructions and screengrabs on 24 May 2023.*
*Updated testing instructions with new Gutenberg Experiments setting on 30 May 2023.*

## What?

### This pull request is a replacement for #49621, which was closed by Github's UI due to the target branch being merged to trunk. We can continue the conversation here.

This is part of the experiments section with the interactivity API in revising core blocks.
It uses directives from the [Directives hydration system](https://github.com/WordPress/block-interactivity-experiments/issues/80)

## Why?
It would enable native support for image lightboxes, a commonly used feature in many websites, without the use of jQuery or other external libraries.

## How?
It adds directives on click to images, presenting a larger version of the image along with an overlay, which can be hidden with an additional click, by scrolling, or pressing the escape key.

## Testing Instructions
1. in the admin, enable Gutenberg > Experiments > Interactivity API and Behaviors UI
2. Add an image to a post
3. Test enabling and disabling the lightbox using image's Advanced panel
4. View the post
5. Click on the image

### Testing Instructions for Keyboard
<!-- How can you test the changes by using the keyboard only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->
1. Tab to the image
2. Press Enter to open the lightbox
3. Test pressing Escape to close the lightbox; also test pressing Enter on the close button

## Notes
- This is an early implementation; the idea is for themes to be able to enable or disable the lightbox behavior, for child themes to be able to override that behavior, and for users to be able to toggle lightboxes for all images sitewide or at the block level.
- This is part of a broader initiative to enable adding Interactivity API behaviors to components using the editor.
- Currently, we can just toggle the lightbox for a particular image at the block level under the Advanced panel (see below)

## Screenshots or screencast <!-- if applicable -->

### Editor

https://github.com/WordPress/gutenberg/assets/5360536/9a56a63a-787a-40fc-a622-315847393e33

### Frontend

https://github.com/WordPress/gutenberg/assets/5360536/5f49202d-18d9-4e63-b2e8-daf6888ab2a7

artemiomorales committed:
Add lightbox to image block

First pass at adding lightbox.

Note: Added custom implementation of Preact Portal
because the children[0] declaration in the render method
was erroneous and undefined, preventing it from working
as expected.

artemiomorales committed:
Add logic for hiding lightbox on esc key press and overlay click

artemiomorales committed:
Improve styles and add note to add conditional for lightbox markup

artemiomorales committed:
Add editor UI and attribute for toggling lightbox

artemiomorales committed:
Remove image translation animation and add fade instead

michalczaplinski committed:
Add behaviors to the core theme.json

michalczaplinski committed:
Add behaviors to json schemas for theme.json

artemiomorales committed:
Add accessibility; clean up styles; fix bug regarding ref in directives

Cleaned up the PHP file a bit as I implemented standard accessibility
for a lightbox.

I changed names of attributes and properties to be clearer, as well as
conslidated some CSS.

Important to note: I fixed a bug with the directives wherein 'ref'
was coming in as null due to how it was being passed in hooks.js.

artemiomorales committed:
Configure image to use new Interactivity API runtime included in Gutenberg

artemiomorales committed:
Remove viewScript from image config

artemiomorales committed:
Add Portal directive to Interactivity API runtime

artemiomorales committed:
Set scrim to site background color

artemiomorales committed:
Remove extraneous image CSS declaration

artemiomorales committed:
Improve aria labeling

artemiomorales committed:
Code cleanup; simplify syntax, consolidate code

artemiomorales committed:
Refactor code, remove event listeners, consolidate logic

artemiomorales committed:
Fix formatting in SCSS file

artemiomorales committed:
Change CheckboxControl to a ToggleControl; update API docs

artemiomorales committed:
Update wp_enqueue_script to correctly add interactivity runtime

michalczaplinski committed:
Add a behaviors panel

artemiomorales committed:
Fix linter errors

artemiomorales committed:
Update to use core.image namespace

artemiomorales committed:
Pause closing of lightbox slightly when using the mousewheel

artemiomorales committed:
Rename portal directive to 'wp-body' and remove unused reference

artemiomorales committed:
Add internal dependencies flag; update comment

artemiomorales committed:
Remove extraneous code

michalczaplinski committed:
Remove the changes to theme.json schema

michalczaplinski committed:
Add behaviors to the VALID_SETTINGS in class-wp-theme-json-gutenberg.php

michalczaplinski committed:
Add the first (still broken) version of the lightbox settings

artemiomorales committed:
Fix accessibility bug due to directives being defined incorrectly

artemiomorales commented on packages/block-library/src/image/interactivity.js:
```diff
@@ -0,0 +1,84 @@
+
+const raf = window.requestAnimationFrame;
+// Until useSignalEffects is fixed: https://github.com/preactjs/signals/issues/228
+const tick = () => new Promise( ( r ) => raf( () => raf( r ) ) );
```
@SantosGuillamot I rebased on trunk and the issue still exists; it doesn't work without the tick.

SantosGuillamot replied:
I've been taking a deeper look at this and you're right the issue still exists. Sometimes it gets the correct focus and sometimes it doesn't, so I assume it is a race condition or something similar.

Exploring it deeper, it seems the issue is caused by the use of `visibility: hidden/visible` in [the CSS animations](https://github.com/artemiomorales/gutenberg/blob/experiment/image-with-interactivity-api/packages/block-library/src/image/style.scss#L239-L267). Removing that part, and moving it to the active class selector, seems to fix the problem (at least for now). I made a quick video explaining what I saw:

https://www.loom.com/share/63bb6d517f2a47398101d15be6b70db6

artemiomorales replied:
Thanks for the feedback!

Visibility is is in the animation to ensure the overlay doesn't disappear before the opacity reaches 0 in the fade out animation.

That being said, I realized that we actually don't need that during the fade in animation, and since that's the only time we have the focus problem, we can remove the visibility attribute in that case.

I've gone ahead and made that change, as well as removed the `tick()` 🙌

artemiomorales committed:
Move enableLightbox declaration in block.json

artemiomorales committed:
Remove extraneous package.json declarations

artemiomorales committed:
Fix PHP indentation linter error

artemiomorales committed:
Revise package-lock.json to match trunk

artemiomorales committed:
Fix linter errors in portals.js

artemiomorales committed:
Update fixtures

artemiomorales commented on packages/block-library/src/image/block.json:
```diff
@@ -80,6 +80,10 @@
+		},
+		"enableLightbox": {
+			"type": "boolean",
+			"default": true
```
@SantosGuillamot You were right. I initially didn't see 'enableLightbox' in the block's saved markup and thought it was broken; I then realized that the block only saves the attribute if it differs from the default value. This has been fixed. Thanks!

Context: This is in response to a [comment](https://github.com/WordPress/gutenberg/pull/49621/files#r1185769629) on the old pull request.

michalczaplinski committed:
WIP: Added a SelectControl for the behaviors

gziolo committed:
Tests: Add e2e specs for Image block on the frontend

michalczaplinski committed:
Format PHP

michalczaplinski committed:
Format correctly again

michalczaplinski committed:
Use the props.name when getting the behaviors

gziolo commented:
I paired with Artemio yesterday, and I wanted to reshare some high-level comments:
- I think the lightbox should be disabled by default for a start, the user would need to explicitly enable it, or with the work on Block Behaviors (https://github.com/WordPress/gutenberg/pull/49972), it’s the theme author who could enable it for all Image blocks with a single setting.
- We should consider putting the lighbox behind the feature flag that needs to be enabled on the Gutenberg plugin settings page. We could reuse the same flag as the Navigation block (https://github.com/WordPress/gutenberg/pull/50041). The reason is that they both use Interactivity API, so it would be beneficial to have a shared way to control whether the experience is present on the site or not.

Otherwise, I think that this PR is shaping up nicely, and we should start testing the logic to plan for landing it in the Gutenberg plugin.

artemiomorales committed:
Disable lightbox by default

artemiomorales committed:
Add aria-role attribute to lightbox markup

cbravobernal committed:
Add initial e2e tests

michalczaplinski committed:
Update the withBehaviors description

michalczaplinski committed:
Cleaned up behaviors.js

artemiomorales committed:
Add tests

artemiomorales committed:
Update fixtures

artemiomorales commented:
@gziolo I've just pushed up the tests.
The two tests potentially missing are 1.) tabbing to the lightbox and 2.) having multiple lightboxes on the page at once.

I'm not sure if the first is necessary and haven't figured out a way to tab through the page elements and compare locators to determine if the lightbox button is selected.

The second would be nice to cover, and I'll get to it when I can.

For the moment, syncing this with [WIP: Behaviors UI](https://github.com/WordPress/gutenberg/pull/49972) seems more important than either of these tests, so I'm going to start looking at that.

gziolo commented on packages/block-library/src/image/interactivity.js:
```diff
@@ -0,0 +1,84 @@
+							event.preventDefault();
+						}
+
+						if ( escapeKeyPressed || isTabKeyPressed ) {
```
What's the rationale behind using a Tab key to close the lightbox?

The handling is for the container with the `dialog` role, so I assumed that you the Escape key would trigger that, or pressing Enter or Space on the close button.

artemiomorales replied:
This comes from the original [set of recommendations](https://github.com/WordPress/gutenberg/pull/49621#issuecomment-1508439898) you posted in the initial PR. 

> Either of these:
> a. Tabbing beyond the last link in the lightbox or shift-tabbing before the first link should close the lightbox
> b. Tabbing should cycle within the actionable links within the lightbox until user clicks a close link.
> 

Since we only have one actionable item inside the lightbox, the close button, it made sense to me to go with option A rather than create a focus trap.

Also, rather than tabbing to some unspecified next link on the page, I thought it made sense to put focus on the enable lightbox button when the lightbox closes, that way folks don't lose their place.

joedolson replied:
+1 to @gziolo's comment. Only Esc, Enter, and Space should close the dialog.

cbravobernal committed:
Update e2e tests to check visibility

cbravobernal committed:
Update core-blocks doc with behaviors attribute

andrewhayward commented:
I wanted to make sure that @joedolson and @afercia's [accessibility comments](https://github.com/WordPress/gutenberg/pull/49621#issuecomment-1525573008) didn't get lost in the move, and mirror their thoughts around modal dialog interaction.

To summarise, and [as previously stated](https://github.com/WordPress/gutenberg/pull/49621#issuecomment-1509040080), while the interaction might seem simple and straightforward, and the user experience not that of a typical modal dialog, semantically it essentially just that. Which means we should be careful to provide appropriate functionality.

Any trigger that opens the image overlay should (semantically) present as a button, and be labelled appropriately to explain what it does. This means we cannot use the image itself as the trigger, because that should (semantically) present as an image, and have a label reflecting what it actually is ('alt text').

We could hide the trigger button until it receives focus, or the image is hovered, but that brings its own discovery issues and cognitive accessibility questions. Ideally it would be present at all times.

When open, everything behind it should be inert. There should be an obvious way to close the overlay (typically an `×` button), as well as respecting the `esc` key. Generally, tapping the overlay background closes it too. Focus should return to the element that opened it; most likely the original trigger button.

---

I'm not sure how much of that has been covered, but [comments in the previous PR](https://github.com/WordPress/gutenberg/pull/49621#issuecomment-1531143049) would suggest some of it is already tackled. The most important thing to remember is that while it might visually be a pretty lightweight experience, we should still make sure the appropriate controls and affordances are in place to cover different access needs.

joedolson commented on packages/block-library/src/image/index.php:
```diff
@@ -54,6 +54,7 @@ function render_block_core_image( $attributes, $content ) {
 							$content
 						</button>
 						<div data-wp-body="" class="wp-lightbox-overlay"
+							aria-role="dialog"
```
The correct role is `role="dialog"`; aria-role isn't a valid attribute.

joedolson commented on packages/block-library/src/image/index.php:
```diff
@@ -27,16 +28,67 @@ function render_block_core_image( $attributes, $content ) {
+							data-wp-on.click="actions.core.image.hideLightbox"
+							>
+								$content
+								<button aria-label="Close lightbox" class="close-button" data-wp-on.click="actions.core.image.hideLightbox">
```
Placing the button after the image is a confusing UI for screen readers: the close button should be the first thing inside the dialog, so the next content is the image itself. Right now, the user would need to navigate backwards to examine the image.

joedolson commented on packages/block-library/src/image/index.php:
```diff
@@ -27,16 +28,67 @@ function render_block_core_image( $attributes, $content ) {
+					data-wp-context='{ "core": { "image": { "initialized": false, "lightboxEnabled": false, "lastFocusedElement": null } } }'>
+						<button aria-haspopup="dialog" aria-label="$aria_label" data-wp-on.click="actions.core.image.showLightbox">
+							$content
+						</button>
```
Image captions can contain links; with this model, those links would be nested inside a button, which is a) not allowed and b) creates really unpredictable user interactions.

This needs some further working out to cover scenarios like captions containing links and images that are linked.

artemiomorales replied:
Ok I've updated the implementation so the image is no longer inside of a button, so that should cover the caption scenario.

As for the image being a link, the lightbox functionality will be overriden and its markup won't render if the user configures the image to be a link.

joedolson commented on packages/block-library/src/image/index.php:
```diff
@@ -27,16 +28,67 @@ function render_block_core_image( $attributes, $content ) {
+							data-wp-on.keydown="actions.core.image.handleKeydown"
+							data-wp-on.mousewheel="actions.core.image.hideLightbox"
+							data-wp-on.click="actions.core.image.hideLightbox"
+							>
```
The dialog should have a label of some kind. Normally this is conveyed via aria-labelledby pointing to content inside the modal; e.g., the image itself.

artemiomorales replied:
Ok this has been updated

joedolson commented on packages/block-library/src/image/interactivity.js:
```diff
@@ -0,0 +1,84 @@
+							( event.type === 'click' &&
+								event.pointerType === '' )
+						) {
+							context.core.image.lastFocusedElement.focus();
```
In my testing, this is failing for the Enter and Space keys when a screen reader is running; the focus is set correctly using Esc and Tab (although tab shouldn't close the dialog.) This is probably because the screen reader is intercepting keypresses at some level. Tested with NVDA/Chrome and NVDA/Firefox. Focus is being lost.

SantosGuillamot replied:
In which case we don't want to focus the `lastFocusedElement` when the lightbox is hidden? Couldn't we just move the `context.core.image.lastFocusedElement.focus();` outside of the conditional?

SantosGuillamot replied:
I was just testing it and removing the conditional, and just calling `context.core.image.lastFocusedElement.focus()`,  seems to solve the issue with NVDA. It works with Enter and Space as well.

Don't we want to focus the last element when the lightbox is hidden after scrolling or when clicking outside?

artemiomorales replied:
> Don't we want to focus the last element when the lightbox is hidden after scrolling or when clicking outside?

@SantosGuillamot Thanks for looking into this! I previously didn't take this approach because setting the focus when closing the dialog with a mouse click produced a visible focus outline, which would have been distracting for mouse users. However, in the latest push, I moved the image outside of the `button` element, which fortunately also resolves the UI clutter, so I've been able to remove this conditional 👍

joedolson reviewed (changes requested):
Thanks for working on this! I've flagged a number of continuing accessibility issues that'll need to be handled before this can move forward.

cbravobernal committed:
Update fixtures with new lightbox attribute

michalczaplinski committed:
Add a new theme for e2e tests

michalczaplinski committed:
Change theme.json to include `settings.behaviors`

michalczaplinski committed:
Remove default behavior value from block.json

michalczaplinski committed:
Minor fix to the implementation

michalczaplinski committed:
Add more e2e tests

draganescu commented:
Hi @andrewhayward - I am no accessibility expert just a dev who reading this bit:

> We could hide the trigger button until it receives focus, or the image is hovered, but that brings its own discovery issues and cognitive accessibility questions.

... felt a bit caught off guard. The fact that "hide the trigger button until it receives focus" which we do all over the place via a `VisuallyHidden` component is problematic for cognitive accessibility needs raises a question to me:

- how much of web accessibility is making web content (be it document or application) available to assisitive technology users and how much of the scope of all accessibility needs are the responsibility of the assistive technology itself to handle? 

In particular, using `VisuallyHidden` elements that are semantic and focusable, just not visible, should be something that the assisitive technology used by the user should handle for cognitive accessibility needs and describe the semantics - which exist - in a way that is predictable and actionable including in cognitive accessibility needs contexts. Is this a wrong line of thinking?

michalczaplinski committed:
Create a selector for behaviors

michalczaplinski committed:
Add a filter to load behaviors on the server from the theme.json

michalczaplinski committed:
Add behaviors on the top-level in core theme.json

michalczaplinski committed:
Use the behaviors in the hooks

michalczaplinski committed:
Update the comment

michalczaplinski committed:
Update the comment in behaviors.js

andrewhayward commented:
Thanks for follow-up, @draganescu. Great questions.

> How much of web accessibility is making web content (be it document or application) available to assistive technology users and how much of the scope of all accessibility needs are the responsibility of the assistive technology itself to handle?
> 
> In particular, using `VisuallyHidden` elements that are semantic and focusable, just not visible, should be something that the assistive technology used by the user should handle for cognitive accessibility needs and describe the semantics - which exist - in a way that is predictable and actionable including in cognitive accessibility needs contexts. Is this a wrong line of thinking?

Very often when we think about accessibility, we consider the technologies that users might make use of to access content, but fall short of thinking about the users themselves. [Cognitive accessibility](https://www.w3.org/WAI/WCAG2/supplemental/#-cognitive-accessibility-guidance) falls into a murky area between traditional accessibility standards and UX.

There's a lot to go into that would be a whole discussion in its own right, but the biggest takeaway in this context is that not everyone who has particular access needs meets them with assistive technology. So your line of thinking isn't necessarily wrong, just maybe not broad enough either. We cannot know who is accessing content, or how, so we have a responsibility to make it as clear and straightforward as we can, both semantically and visually.

gziolo commented:
With https://github.com/WordPress/gutenberg/commit/6f2d112399b007ccf33e913183e36446d841c9f7, there might be slight changes necessary in the webpack config and the PHP file that register the view script. The built file is going to be located in `build/block-library/blocks/image/interactivity.min.js`.

cbravobernal committed:
Update fixtures and e2e

artemiomorales committed:
Fix accessibility issues

1.) Changed 'aria-role' attribute to 'role'
2.) Moved lightbox close button to before modal content
3.) Remove logic for closing dialog with Tab key
4.) Move <img> outside of <button> element
5.) Add aria-labelledby attribute to dialog

cbravobernal committed:
Prevent mobile test gutenberg error (temporary - not the best solution)

cbravobernal committed:
Add priority on filter

cbravobernal committed:
Fix php standards

cbravobernal committed:
Found a much better way to fix php tests

cbravobernal committed:
Small refactor

michalczaplinski committed:
Add `behaviors` as an allowed key to BLOCK_EDITOR_SETTINGS

michalczaplinski committed:
Move the behaviors to top level in the e2e test theme.json file

michalczaplinski committed:
Rename the `behaviors` setting to `behaviorsUIEnabled`

michalczaplinski committed:
Change "None" to "No behaviors"

michalczaplinski committed:
Behaviors ->  behavior

michalczaplinski committed:
Fix redundant ternary

michalczaplinski committed:
Improve the JSDoc for behaviors selector

michalczaplinski committed:
Rename the test themes to make more sense

artemiomorales committed:
Add focus trap

artemiomorales committed:
Label elements as 'inert' when lightbox is open

artemiomorales committed:
Update import of image interactivity scripts

artemiomorales committed:
Fix bug where interactivity only worked with admin enabled

artemiomorales committed:
Put lightbox rendering behind experimental flag

artemiomorales commented:
I've pushed a batch of changes that covers @joedolson's accessibility feedback. I've marked 'resolved' for the two simplest issues; for the rest, I see them as handled on my end but believe this PR is ready for another review.

> Any trigger that opens the image overlay should (semantically) present as a button, and be labelled appropriately to explain what it does. This means we cannot use the image itself as the trigger, because that should (semantically) present as an image, and have a label reflecting what it actually is ('alt text').

This feedback from @andrewhayward has also been addressed, namely the image itself is no longer the trigger — instead the button is overlaid on top of the image. The button also has an aria-label, though does not have a visible label in the UI pending additional explorations and discussion with design folks.

Also, I just realized a few of my comments from several days ago never became public due to a snafu on my end with the Github UI. That's now been updated, so @gziolo and @SantosGuillamot should see responses to their questions from a while back. Sorry for the delay on that!

**Important: I've also rebased on trunk and the lightbox is now hidden behind an experimental flag. To activate it, enable Gutenberg > Experiments > Core blocks in the admin.**

SantosGuillamot commented:
On another note, I was experimenting with the possibility of using a zooming animation for the lightbox, just in case we want to consider it. The idea was to create a similar effect to the UX of the [View Transitions API](https://developer.chrome.com/docs/web-platform/view-transitions/). This is how the experiment looks like at this moment (I did it before the latest changes where @artemiomorales changed the trigger) :

### Desktop

https://github.com/WordPress/gutenberg/assets/34552881/21e9aa36-3352-4b9c-9d06-50fdc821e205

### Mobile

https://github.com/WordPress/gutenberg/assets/34552881/099fb143-1ab2-462e-9add-086cb032a6ff

### Explanation

_EDIT: I changed the code to use CSS variables which makes the implementation easier, so this video is a bit outdated._ 

The code of that experiment can be found in [this branch](https://github.com/artemiomorales/gutenberg/tree/experiment/add-zoom-animation), and I also made a video explaining it a bit.

https://www.loom.com/share/b7138ad47a6a42adb7808f3b07a193e4

If at the end we want to follow this approach, I'm happy to help polish the code 🙂

michalczaplinski committed:
Remove definition of `behaviors` attribute in core/image

michalczaplinski committed:
Change default value for `behaviors.lightbox` to false and update e2e tests accordingly

michalczaplinski committed:
Change the way we get the data from `theme.json` and adjust e2e

michalczaplinski committed:
Capitalize behaviors' labels

michalczaplinski committed:
Move PHP code adding `theme.json` behaviors to `block-editor-settings`

michalczaplinski committed:
Update comment

michalczaplinski committed:
Remove the behaviors require from load.php

michalczaplinski committed:
Revert "Update comment"

This reverts commit fc812bc6c04e8b3c5c7fe3450c1e39d1dd6f6f3b.

michalczaplinski committed:
Revert "Move PHP code adding `theme.json` behaviors to `block-editor-settings`"

This reverts commit e8d16c8bb4977d1e27b758a067066a78fc21f794.

artemiomorales committed:
Add check for query selector and update tests

artemiomorales committed:
Update spacing in PHP file

SantosGuillamot commented on packages/block-library/src/image/interactivity.js:
```diff
@@ -0,0 +1,145 @@
+					context.core.image.siteBlocksElement =
+						document.querySelector( '.wp-site-blocks' );
+
+					context.core.image.documentElement.classList.add(
+						'has-lightbox-open'
+					);
```
What is the point of storing `document.documentElement` in the context? If we are just using it here, couldn't we do directly `document.documentElement.classList.add( 'has-lightbox-open')?

artemiomorales committed:
Remove tick() and modify CSS

SantosGuillamot commented on packages/block-library/src/image/interactivity.js:
```diff
@@ -0,0 +1,145 @@
+						'has-lightbox-open'
+					);
+
+					if ( context.core.image.adminElement ) {
+						context.core.image.adminElement.setAttribute(
+							'inert',
+							true
+						);
+					}
+
+					if ( context.core.image.siteBlocksElement ) {
+						context.core.image.siteBlocksElement.setAttribute(
+							'inert',
+							true
+						);
+					}
```
Is it needed to differentiate between admin and not admin? I don't like too much sending code only relevant to the admin for all the page visits.

Sorry if it doesn't make sense but, can't we just add the `inert` property to the body in both cases?

artemiomorales replied:
> Is it needed to differentiate between admin and not admin? I don't like too much sending code only relevant to the admin for all the page visits.

Nope, it's done this way because I'm trying to target the two elements separately.

> Sorry if it doesn't make sense but, can't we just add the inert property to the body in both cases?

We can't, because the lightbox is a child of the body and will also become inert in that case.

I was hoping to get feedback on what would be the best way to mark the appropriate DOM elements as inert, or if it's even necessary if we've already marked the dialog as being open, have trapped focus, and have added a scrim to show the background elements as inert. Also, targeting these two HTML elements doesn't guarantee that other interactive elements won't be on the page. For example, at the moment it's not capturing the 'Skip to Content' button or other elements that may be added dynamically to the DOM.

At the same time, creating logic to loop through children of the body to mark them as inert seemed like overkill.

Perhaps @andrewhayward can offer guidance?

SantosGuillamot replied:
Okay, now I understand 🙂 Thanks for the comment!

I'm far from an expert in accessibility but, as you say, marking the dialog, trapping the focus, and adding a background covering the whole screen, seem enough for me. Actually, the current navigation block is not using inert and it follows a similar approach to this one. Anyway, I'd love to understand if/why inert is needed.

> At the same time, creating logic to loop through children of the body to mark them as inert seemed like overkill.

If we finally have to include `inert` to all the elements that are not the lightbox, it seems safer to include it to all the children of the body that are not the current lightbox, right? Maybe some plugins or themes are adding other elements that we are not considering. Something like this: 

```js
for (const element of document.querySelector("body").children) {
  !(element === context.core.image.currentLightbox) &&
    element.setAttribute("inert", "");
}
```

Anyway, I'd wait to confirm if `inert` is really necessary for this particular use case.

artemiomorales replied:
Given that the lightbox already follows a similar pattern to the navigation block, I've gone ahead and removed the `inert` attribute logic for now pending additional discussions. Since this PR is coming out with the lightbox disabled by default and as part of an experiment, the plan is to continue testing accessibility more widely and continue making improvements, so we can always add this back in.

SantosGuillamot commented on packages/block-library/src/image/index.php:
```diff
@@ -27,16 +28,76 @@ function render_block_core_image( $attributes, $content ) {
+						$body_content
+						<div data-wp-body="" class="wp-lightbox-overlay"
+							role="dialog"
+							aria-hidden="true"
+							aria-modal="false"
```
I think we can remove these because they should be handled by `data-wp-bind.aria-hidden` and `data-wp-bind.aria-modal`.

SantosGuillamot commented on packages/block-library/src/image/index.php:
```diff
@@ -27,16 +28,76 @@ function render_block_core_image( $attributes, $content ) {
+					data-wp-context='{ "core": { "image": { "initialized": false, "lightboxEnabled": false, "lastFocusedElement": null } } }'>
+						$body_content
+						<div data-wp-body="" class="wp-lightbox-overlay"
+							role="dialog"
```
Should we use `data-wp-bind.dialog` to toggle between `role="dialog"` and `role=""`? Something like we are doing in [the navigation block](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-library/src/navigation/interactivity.js#L45-L56).

artemiomorales replied:
Sounds good — I created the selector 👍

SantosGuillamot commented on packages/block-library/src/image/block.json:
```diff
@@ -80,6 +80,10 @@
 			"selector": "figure > a",
 			"attribute": "target"
+		},
+		"enableLightbox": {
```
Do you know if using the `enableX` is used somewhere else to define an attribute name? From what I can see attributes usually don't include that syntax. Should we just use `lightbox`?

gziolo replied:
In https://github.com/WordPress/gutenberg/pull/49972 it's a `lightbox` field inside the `behaviors` object. Let's align both implementations.

SantosGuillamot commented on packages/block-library/src/image/index.php:
```diff
@@ -27,16 +28,76 @@ function render_block_core_image( $attributes, $content ) {
+			<<<HTML
+				<div class="wp-lightbox-container"
+					data-wp-island=""
+					data-wp-context='{ "core": { "image": { "initialized": false, "lightboxEnabled": false, "lastFocusedElement": null } } }'>
```
If I am not mistaken, we don't need to define a property that is null here. Maybe we can remove it?

SantosGuillamot reviewed (commented):
I left some comments so far, although I'd like to do a deeper review later today. Thanks for taking care of it, Artemio 🙂

gziolo commented on packages/block-library/src/image/index.php:
```diff
@@ -27,16 +28,76 @@ function render_block_core_image( $attributes, $content ) {
+	$link_destination = isset( $attributes['linkDestination'] ) ? $attributes['linkDestination'] : 'none';
+	$enable_lightbox  = isset( $attributes['enableLightbox'] ) ? $attributes['enableLightbox'] : false;
+
+	if ( gutenberg_is_experiment_enabled( 'gutenberg-interactivity-api-core-blocks' ) && 'none' === $link_destination && $enable_lightbox ) {
```
Let's ensure `gutenberg_is_experiment_enabled` is expressed using only native WordPress APIs because this file gets copied as is to WordPress core.

gziolo replied:
In effect, it's enough use the code from the function:
https://github.com/WordPress/gutenberg/blob/97ab6b167cc0d9da6c5f0ecf6be9457570d33731/lib/load.php#L26-L29

artemiomorales replied:
Thanks, fixed!

artemiomorales committed:
Update config based on behaviors UI and trunk changes

artemiomorales committed:
Remove extraneous code

artemiomorales commented:
> On another note, I was experimenting with the possibility of using a zooming animation for the lightbox, just in case we want to consider it. The idea was to create a similar effect to the UX of the [View Transitions API](https://developer.chrome.com/docs/web-platform/view-transitions/). This is how the experiment looks like at this moment (I did it before the latest changes where @artemiomorales changed the trigger) :
> 

Thanks for this exploration @SantosGuillamot ! This looks good, though I think what's best is to go ahead with the simple fade for now, and we'll add this zoom in a follow up PR that also adds a toggle so folks can opt for their preferred animation

SantosGuillamot commented on packages/block-library/src/image/interactivity.js:
```diff
@@ -0,0 +1,138 @@
+		core: {
+			image: {
+				initLightbox: async ( { context, ref } ) => {
+					const focusableElements =
+						ref.querySelectorAll( focusableSelectors );
+					context.core.image.firstFocusableElement =
+						focusableElements[ 0 ];
+					context.core.image.lastFocusableElement =
+						focusableElements[ focusableElements.length - 1 ];
```
Should we move this inside the `if ( context.core.image.lightboxEnabled )` conditional? We only want to initialize them when the lightbox is opened, right?

artemiomorales replied:
Sure thing, this has been changed 👍

SantosGuillamot commented on test/integration/fixtures/blocks/core__image__attachment-link.json:
```diff
@@ -7,7 +7,11 @@
 			"href": "http://localhost:8888/?attachment_id=7",
-			"linkDestination": "attachment"
+			"linkDestination": "attachment",
+			"behaviors": {
+				"lightbox": true
+			},
+			"lightbox": false
```
This is a question for all the tests: Why do we need `behaviors.lightbox = true` and `lightbox = false`?

artemiomorales replied:
I just touched base with @michalczaplinski on this; there's some inconsistency between this PR and the [behaviors UI PR](https://github.com/WordPress/gutenberg/pull/49972) that he and I will work to resolve between today and early tomorrow, so that should fix this.

michalczaplinski replied:
I've removed it as it's not needed for the Behaviors UI 👍

artemiomorales committed:
Remove extraneous code and simplify variable name

artemiomorales committed:
Use core WordPress API for experiments flag

artemiomorales committed:
Clean up declarations

artemiomorales committed:
Fix PHP spacing

michalczaplinski committed:
Remove the comment that was added previously

michalczaplinski committed:
Update comments in `behaviors.php`

michalczaplinski committed:
Add back the require_once in load.php

michalczaplinski committed:
Use `settings.blocks.core/image.behaviors.lightbox`

michalczaplinski committed:
Use `behaviors.blocks.core/image.lightbox`

michalczaplinski committed:
Remove `enableLightbox`

michalczaplinski committed:
Remove defaultValue from image block.json

michalczaplinski committed:
Update render_callback of core/image to use Behaviors UI

michalczaplinski committed:
Fix e2e test:
- Remove trailing `}`
- Incorrectly assuming `"lightbox": true`

michalczaplinski commented:
Hey @artemiomorales 

Feel free to look at the commits that I pushed. I've fixed the integration with the Behaviors UI, removed the `enableLightbox` attribute, removed the `defaultValue` from the `block.json`, and fixed an e2e test.

I've tested it locally and found no issues.

I think we should consider removing the "experiments" flag like mentioned by @luisherranz  As the lightbox behavior is disabled by default and the user has to select it via dropdown, we could remove the flag.
   <img width="585" alt="Screenshot 2023-05-23 at 21 38 35" src="https://github.com/WordPress/gutenberg/assets/5417266/4338acb7-50dc-4380-864d-4008491a4aa6">

artemiomorales committed:
Remove 'inert' declarations for now

artemiomorales committed:
Add period to behaviors help text

artemiomorales committed:
Create selector for 'role' attribute

artemiomorales committed:
Update fixtures

artemiomorales committed:
Update comment

SantosGuillamot commented on packages/block-library/src/image/interactivity.js:
```diff
@@ -0,0 +1,110 @@
+					context.core.image.lightboxEnabled = true;
+					context.core.image.lastFocusedElement =
+						window.document.activeElement;
+					context.core.image.scrollPosition = window.scrollY;
```
If I am not mistaken, we are missing here the code to add the `has-lightbox-open` class to the `<html>` element, right? I mean this:

```js
document.documentElement.classList.add(
    'has-lightbox-open'
);
```

artemiomorales replied:
Right, this was accidentally removed in a recent push. It's been put back.

artemiomorales commented:
> I think we should consider removing the "experiments" flag like mentioned by @luisherranz As the lightbox behavior is disabled by default and the user has to select it via dropdown, we could remove the flag.

@michalczaplinski Thanks for doing the integration! I've tried removing the "experiments" flag but the lightbox won't work without the interactivity scripts that are only included if the experiments are enabled. Ideally I think we should also put the behaviors UI behind an experimental flag for the best experience.

Other than that, I've cleaned up the tests and done a few other miscellaneous fixes.

SantosGuillamot commented on packages/block-library/src/image/index.php:
```diff
@@ -27,16 +28,88 @@ function render_block_core_image( $attributes, $content ) {
+		return
+			<<<HTML
+				<div class="wp-lightbox-container"
+					data-wp-island=""
```
Not a big deal, but I believe we could use `data-wp-island` instead of `data-wp-island=""` because it is a [boolean attribute](https://developer.mozilla.org/en-US/docs/Glossary/Boolean/HTML).

artemiomorales commented:
Hi all, I believe this PR is ready to go for the purposes of this experiment around the Interactivity API and behaviors.

We still have some outstanding discussion on accessibility, but I think merging this in the next Gutenberg release would allow us to start testing more widely, and we can continue making improvements to accessibility and the design while involving more folks in the discussion.

We're also working on hiding the dropdown from the [behaviors UI pull request](https://github.com/WordPress/gutenberg/pull/49972), which @c4rl0sbr4v0 is looking at.

Other outstanding items include:
-Adding automated tests for tabbing through caption links in the lightbox (this requires adding support for captions in the `ImageBlockUtils` in the `image.spec.js`)
-Adding the zoom animation and associated UI for toggling between different lightbox animations

kevin940726 commented on test/e2e/specs/editor/blocks/image.spec.js:
```diff
@@ -729,6 +729,190 @@ test.describe( 'Image', () => {
+	} );
+
+	test.beforeEach( async ( { admin, page, editor, imageBlockUtils } ) => {
+		await admin.visitAdminPage(
+			'/admin.php',
+			'page=gutenberg-experiments'
+		);
+
+		await page
+			.locator( `#gutenberg-interactivity-api-core-blocks` )
+			.setChecked( true );
+
+		await page.locator( `input[name="submit"]` ).click();
+		await page.waitForLoadState();
```
It would be best if we could implement this as a `requestUtils`, so that we can speed up the test by just calling `fetch` for preparation.

There might not be an official REST API for enabling gutenberg experiments, so we might need to creative for that 😅 . It should not be difficult to do though.

artemiomorales replied:
We're currently discussing the experimental flag and it looks like it's going to be removed, so we'll be able to remove these lines as well @luisherranz @c4rl0sbr4v0

kevin940726 replied:
It can be a follow-up though and I can help with that! We can just add a comment above in case we forget 😅 .

artemiomorales committed:
Clean up data-wp-island attribute

kevin940726 commented on test/e2e/specs/editor/blocks/image.spec.js:
```diff
@@ -729,6 +729,190 @@ test.describe( 'Image', () => {
+<figure class="wp-block-image size-full"><img src="[^"]+\\/${ filename }\\.png" alt="" class="wp-image-\\1"/></figure>
+<!-- \\/wp:image -->`
+		);
+		expect( await editor.getEditedPostContent() ).toMatch( testForTrue );
```
Could we instead use `editor.getBlocks` to improve readability? See this [doc](https://github.com/WordPress/gutenberg/blob/trunk/docs/contributors/code/e2e/overusing-snapshots.md) for more info.

kevin940726 commented on test/e2e/specs/editor/blocks/image.spec.js:
```diff
@@ -729,6 +729,190 @@ test.describe( 'Image', () => {
+		page,
+	} ) => {
+		await page.getByRole( 'button', { name: 'Advanced' } ).click();
+		await page.getByLabel( 'Behaviors' ).selectOption( 'Lightbox' );
```
Nit: Normally `getByRole` is preferred whenever possible, it helps the readers to understand what `role` we are interacting with.

artemiomorales replied:
I've tried using `getByRole` here without success. I've tried various different forms of the following syntax, but it doesn't find the select:

```
const select = page.getByRole( 'select' ).filter( {
	has: page.getByRole( 'option', { value: 'lightbox' } ),
} );
await select.selectOption( 'lightbox' );
```

kevin940726 replied:
The role is likely `combobox` rather than `select`.

Try something like:

```js
page.getByRole( 'region', { name: 'Editor settings' } )
  .getByRole( 'combobox', { name: 'Behaviors' } )
```

artemiomorales replied:
Perfect, thanks! Searching for `combobox` worked.

kevin940726 commented on test/e2e/specs/editor/blocks/image.spec.js:
```diff
@@ -729,6 +729,190 @@ test.describe( 'Image', () => {
+			openLightboxButton = page.getByRole( 'button', {
+				name: 'Open image lightbox',
+			} );
+			lightbox = page.locator( '.wp-lightbox-overlay' );
```
Could this be refactored into `getByRole`? If it's not accessible, could we add a `data-testid`? If it's not desirable, which I understand for frontend code, could we add a simple comment above here for readers?

artemiomorales replied:
Done 👍

kevin940726 commented on test/e2e/specs/editor/blocks/image.spec.js:
```diff
@@ -729,6 +729,190 @@ test.describe( 'Image', () => {
+			await page.getByRole( 'button', { name: 'Advanced' } ).click();
+			await page.getByLabel( 'Behaviors' ).selectOption( 'Lightbox' );
+
+			postId = await editor.publishPost();
```
Nit: I guess we can use `const` here since it's not referenced anywhere else?

artemiomorales replied:
It was used twice, but noted that it's probably clearer to define it within the test it's clearer, so this has been changed.

artemiomorales committed:
Readd mistakenly deleted line

SantosGuillamot reviewed (commented):
If I tested it correctly, it seems we are sending the JavaScript files even if the images on the page don't have the lightbox activated. Is this expected?

artemiomorales committed:
Improve syntax of tests

kevin940726 commented on test/e2e/specs/editor/blocks/image.spec.js:
```diff
@@ -848,15 +848,15 @@ test.describe( 'Image - interactivity', () => {
 				name: 'Open image lightbox',
 			} );
-			lightbox = page.locator( '.wp-lightbox-overlay' );
+			lightbox = page.getByRole( 'dialog' );
```
This might be a concern here that the dialog doesn't have an accessible name? As recommended by the [guideline](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/dialog_role#labeling).

artemiomorales replied:
It doesn't have a generic name by default, but it has an `aria-labelledby` attribute, which refers to its image.

artemiomorales commented on lib/experimental/interactivity-api/blocks.php:
```diff
@@ -230,15 +230,15 @@ function gutenberg_block_core_navigation_add_directives_to_submenu( $w ) {
 function gutenberg_block_update_interactive_view_script( $metadata ) {
 	if (
-		in_array( $metadata['name'], array( 'core/file', 'core/navigation' ), true ) &&
+		in_array( $metadata['name'], array( 'core/file', 'core/navigation', 'core/image' ), true ) &&
```
> If I tested it correctly, it seems we are sending the JavaScript files even if the images on the page don't have the lightbox activated. Is this expected?

@SantosGuillamot I'm following the method for adding the interactivity script as done for the `core/file` and `core/navigation` blocks. Off the top of my head, I'm not sure how we'd detect if the lightbox is being used to prevent these scripts from getting loaded at this level.

SantosGuillamot replied:
I am not familiar with the scripts enqueuing. Maybe @gziolo can help here better. The two solutions that come to my mind are:

* Remove the `core/image` here and enqueue it conditionally in the `index.php` of the image. When [this condition](https://github.com/artemiomorales/gutenberg/blob/experiment/image-with-interactivity-api/packages/block-library/src/image/index.php#L50) is true.
* Dequeue the script using something like [wp_dequeue_script](https://developer.wordpress.org/reference/functions/wp_dequeue_script/) when that condition is false.

gziolo replied:
The plan is to optimize it later. You need to do it globally because one Image block could use the lightbox, and another might not. That's why you can't decide on a single-block level. As of today, when `viewScript` is defined for the block in `block.json` then all listed files are always enqueued, and that's what everyone expects.

In effect, the same applies to the Navigation and File blocks that when rendered always enqueue view scripts even when they aren't necessary.

SantosGuillamot replied:
I believe we could do something like this:

```php
if ( lightbox ) {
    $block->block_type->view_script_handles = array_merge($block->block_type->view_script_handles, array('wp-block-image-view')); 
} else {
    $block->block_type->view_script_handles = array_diff($block->block_type->view_script_handles, array('wp-block-image-view'));
}
```

Anyway, I'll work on that on a follow-up pull request along the Navigation and File blocks. We can resolve this conversation.

artemiomorales committed:
Replace getByLabel() with getByRole() in some tests

cbravobernal committed:
Remove experimental setting for interactivity API

artemiomorales committed:
Remove 'experiments' flag

artemiomorales commented:
For anyone who was just pinged for a review, this PR depends on the [behaviors UI](https://github.com/WordPress/gutenberg/pull/49972) branch and has been rebased on it, so includes code from there as well.

The idea is to merge this one to `trunk` after the behaviors UI has been merged.

artemiomorales committed:
Remove dependency on 'experiments' flag in tests

alexstine commented:
One accessibility note about the trigger button:

```
<button aria-haspopup="dialog" aria-label="Open image lightbox - Testing image." data-wp-on.click="actions.core.image.showLightbox"></button>
```

Instead of " - ALT HERE", let's make it "Open image lightbox: ALT HERE". Using a ":" forces a pause in most screen readers to not read as one continuous string.

Otherwise, this is functioning better with screen readers then I expected.

Thanks.

SantosGuillamot committed:
Revert "Remove dependency on 'experiments' flag in tests"

This reverts commit 429ebbe499fecb2d705764013a53d5d1208ef463.

SantosGuillamot committed:
Revert "Remove 'experiments' flag"

This reverts commit cd0c9dd6d8de6a4066277a155489d65619e8b71e.

SantosGuillamot committed:
Change `aria-label` separator

SantosGuillamot committed:
Add back the experimental setting

SantosGuillamot committed:
Merge branch 'trunk' into experiment/image-with-interactivity-api

carolinan commented on packages/block-library/src/image/index.php:
```diff
@@ -27,16 +28,88 @@ function render_block_core_image( $attributes, $content ) {
+
+	if ( ! empty( $experiments['gutenberg-interactivity-api-core-blocks'] ) && 'none' === $link_destination && $lightbox ) {
+
+		$aria_label = 'Open image lightbox';
```
Is there a reason for not making this text translatable?

gziolo replied:
It definitely should be translated 👍

carolinan commented on packages/block-library/src/image/index.php:
```diff
@@ -27,16 +28,88 @@ function render_block_core_image( $attributes, $content ) {
+		return
+			<<<HTML
+				<div class="wp-lightbox-container"
+					data-wp-island
```
Most of these data attributes are self explanatory, but this "island" part is not, I think this could be improved to be easier to read?

gziolo replied:
This one is special and I run into an issue with it as I had to learn it enables Interactivity API runtime for the block.

luisherranz replied:
That's a good point.

_Island_ is a term that refers to a part of the site that is hydrated and therefore becomes interactive.

The idea of an "islands architecture" was first coined by Etsy's frontend architect [Katie Sylor-Miller](https://twitter.com/ksylor) in 2019, and expanded on in [this post](https://jasonformat.com/islands-architecture/) by Jason Miller (Preact's creator). Since then, multiple JavaScript frameworks have adopted it: ([1](https://docs.astro.build/en/concepts/islands/), [2](https://islandjs.dev/en/), [3](https://iles.pages.dev/)...).

Maybe we can change it to `data-wp-interactive`. Would that be easier to read, or do you have any other suggestions?

carolinan replied:
Yes I think `data-wp-interactive` would be clearer.

luisherranz replied:
Thanks, Carolina 🙂 

I opened [this issue](https://github.com/WordPress/block-interactivity-experiments/issues/238) to discuss it.

_For context, the full version of the Interactivity API is still in that repo, but we'll move everything to the Gutenberg repo after the WP 6.3 feature freeze, including the codebase and discussions._
